### PR TITLE
Change Regex class to use PCRE2

### PR DESCRIPTION
**Pull request by bryancall**
Converted Regex class from PCRE to PCRE2

### CODE REVIEW

**Comment by bryancall** [comment]
[approve ci]

**Comment by bryancall** [comment]
[approve ci]

**Comment by bryancall** [comment]
[approve ci autest]

**Comment by bryancall** [comment]
[approve ci autest]

**Comment by bryancall** [comment]
[approve ci autest]

**Comment by bryancall** [comment]
[approve ci rocky]

**Comment by bryancall** [comment]
[approve ci]

**Comment by bryancall** [comment]
[approve ci]

**Review by traeak — CHANGES_REQUESTED** [review]
Try this patch (inspired by cripts):

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 920d15835..10913055e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -258,7 +258,8 @@ if(LibLZMA_FOUND)
 endif()
 
 find_package(PCRE REQUIRED)
-find_package(PCRE2 COMPONENTS 8BIT)
+
+pkg_check_modules(PCRE2 REQUIRED IMPORTED_TARGET libpcre2-8)
 
 include(CheckOpenSSLIsBoringSSL)
 include(CheckOpenSSLIsQuictls)

File: include/tsutil/Regex.h [hunk 1]
```diff
@@ -28,24 +28,64 @@
 #include <vector>
 #include <memory>
 
-#include "swoc/MemSpan.h"
-
-/// Match flags for regular expression evaluation.
+#define PCRE2_CODE_UNIT_WIDTH 8
+#if __has_include(<pcre2/pcre2.h>)
```
Comment (traeak) [inline include/tsutil/Regex.h:32]: Just replace this logic with #include <pcre2.h>

File: src/tsutil/Regex.cc [hunk 2]
```diff
@@ -26,160 +26,281 @@
...
+  void *ptr = ::malloc(size);
+  return ptr;
+}
 
-  if (flags & RE_ANCHORED) {
-    options |= PCRE_ANCHORED;
+//----------------------------------------------------------------------------
+RegexMatches::~RegexMatches()
+{
+  if (_match_data) {
```
Comment (traeak) [inline src/tsutil/Regex.cc:146]: pls check against nullptr

File: src/tsutil/Regex.cc [hunk 3]
```diff
@@ -26,160 +26,281 @@
...
-  }
+  _code      = that._code;
+  that._code = nullptr;
+}
 
-  return captures;
+//----------------------------------------------------------------------------
+Regex::~Regex()
+{
+  if (_code) {
```
Comment (traeak) [inline src/tsutil/Regex.cc:209]: pld check against nullptr

File: src/tsutil/Regex.cc [hunk 4]
```diff
@@ -26,160 +26,281 @@
...
+  return this->compile(pattern, error, erroroffset, flags);
 }
 
+//----------------------------------------------------------------------------
 bool
-Regex::exec(std::string_view const &str, int *ovector, int ovecsize) const
+Regex::compile(std::string_view pattern, std::string &error, int &erroroffset, uint32_t flags)
 {
-  int rv;
+  if (_code) {
```
Comment (traeak) [inline src/tsutil/Regex.cc:228]: pls check against nullptr

**Comment by bryancall** [comment]
Valgring test
```
14:13:05 zeus:(regex_interface)~/dev/apache/trafficserver/build$ ninja && valgrind ./src/tsutil/test_tsutil
[0/2] Re-checking globbed directories...
[152/152] Linking CXX executable src/traffic_server/traffic_server
==1309146== Memcheck, a memory error detector
==1309146== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==1309146== Using Valgrind-3.22.0 and LibVEX; rerun with -h for copyright info
==1309146== Command: ./src/tsutil/test_tsutil
==1309146==
===============================================================================
All tests passed (182 assertions in 8 test cases)

==1309146==
==1309146== HEAP SUMMARY:
==1309146==     in use at exit: 288 bytes in 6 blocks
==1309146==   total heap usage: 2,955 allocs, 2,949 frees, 684,530 bytes allocated
==1309146==
==1309146== LEAK SUMMARY:
==1309146==    definitely lost: 0 bytes in 0 blocks
==1309146==    indirectly lost: 0 bytes in 0 blocks
==1309146==      possibly lost: 0 bytes in 0 blocks
==1309146==    still reachable: 288 bytes in 6 blocks
==1309146==         suppressed: 0 bytes in 0 blocks
==1309146== Rerun with --leak-check=full to see details of leaked memory
==1309146==
==1309146== For lists of detected and suppressed errors, rerun with: -s
==1309146== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

**Comment by bryancall** [comment]
Rebased the PR.  There was a change in `tls_bridge/CMakeLists.txt` that was conflicting.

**Review by traeak — APPROVED** [review]
Changes look fine.

**Comment by bryancall** [comment]
[approve ci autest]

**Comment by bryancall** [comment]
[approve ci autest]

**Comment by bryancall** [comment]
[approve ci autest]
[approve ci autest]

**Comment by bryancall** [comment]
[approve ci]

**Comment by cmcfarlen** [comment]
Cherry-picked to v10.0.x

**Pull request merged**